docs/RFCS/leaseholder_rebalancing.md

```diff
@@ -0,0 +1,136 @@
+- Feature Name: Leaseholder Rebalancing
+- Status: draft
+- Start Date: 2016-10-26
+- Authors: Peter Mattis
+- RFC PR: [#10262](https://github.com/cockroachdb/cockroach/pull/10262)
+- Cockroach Issue: [#9462](https://github.com/cockroachdb/cockroach/issues/9462)
+
+# Summary
+
+Periodically rebalance range leaseholders in order to distribute the
+per-leaseholder work.
+
+# Motivation
+
+The primary motivation of ensuring leaseholders are distributed
+throughout a cluster is to avoid scenarios in which a node is unable
+to rebalance replicas away because of the restriction that we refuse
+to rebalance a replica for which a node is the leaseholder. This
+restriction is present in order to prevent an availability hiccup on
+the range when the leaseholder is removed from it.
+
+It is interesting to note a problematic behavior of the current
+system. The current leaseholder will extend its lease as long as it is
+receiving operations for a range. And when a range is split, the lease
+for the left-hand side of the split is cloned and given to the
+right-hand side of the split. The combined effect is that a newly
+created cluster that has continuous load applied against it will see a
+single node slurp up all of the range leases which causes a severe
+replica imbalance (since we can't rebalance away from the leaseholder)
+as well as a performance bottleneck. We actually see increased
+performance by periodically killing nodes in the cluster.
+
+The second motivation is to more evenly distributed load in a
+cluster. The leaseholder for a range has extra duties when compared to
+a follower: it performs all reads for a range and proposes almost all
+writes. [Proposer evaluated KV](proposer_evaluated_kv.md) will reduce
+the cost of write KV operations on followers exacerbating the
+difference between leaseholders and followers. These extra duties
+impose additional load on the leaseholder making it desirable to
+spread that load throughout a cluster in order to improve performance.
+
+The last motivation is to place the leaseholder for a range near the
+gateway node that is accessing the range in order to minimize network
+RTT. As an obvious example: it is preferable for the leaseholder to be
+in the same datacenter as the gateway node.
+
+# Detailed design
+
+This RFC is intended to address the first two motivations and punt on
+the last one (load-based leaseholder placement). Note that addressing
+the problem of evenly distributing leaseholders across a cluster will
+also address the inability to rebalance a replica away from the
+leaseholder as we'll always have sufficient non-leaseholder replicas
+in order to perform rebalancing.
```

Comment:
After reading the first motivation I understand what is meant here. I think it can be clarified by saying,

`Note that addressing
 the second motivation of evenly distributing leaseholders across a cluster will
 also address the primary motivation of the inability to rebalance a replica away from the
 leaseholder as we'll always have sufficient non-leaseholder replicas
 in order to perform rebalancing.`

Or perhaps naming these motivations